rct229/rules/section23/section23rule16.py

```diff
@@ -0,0 +1,164 @@
+            standard_section="Section G3.1.3.19 Preheat Coils (Systems 5 through 8)",
+            is_primary_rule=True,
+            rmr_context="ruleset_model_instances/0",
+            list_path="$..heating_ventilating_air_conditioning_systems[*]",
+        )
+
+    def is_applicable(self, context, data=None):
+        rmi_b = context.baseline
+        baseline_system_types_dict = get_baseline_system_types(rmi_b)
+
+        return any(
+            [
+                baseline_system_type_compare(system_type, applicable_sys_type, False)
+                for system_type in baseline_system_types_dict.keys()
+                for applicable_sys_type in APPLICABLE_SYS_TYPES
+            ]
+        )
+
+    def create_data(self, context, data):
+        rmi_b = context.baseline
+
+        # set hvac_id with highest zone_design_heating_setpoint
+        hvac_max_zone_setpoint_dict = {}
+        for zone in find_all("$..zones[*]", rmi_b):
+            zone_design_heating_setpoint = getattr_(
+                zone, "design_heating_setpoint", "design_thermostat_heating_setpoint"
+            )
+            for terminal in find_all("$..terminals[*]", zone):
+                hvac_id = getattr_(
+                    terminal,
+                    "system",
+                    "served_by_heating_ventilating_air_conditioning_system",
+                )
+                hvac_max_zone_setpoint_dict = {
+                    hvac_id: max(
+                        zone_design_heating_setpoint,
+                        hvac_max_zone_setpoint_dict.get(hvac_id, ZERO.TEMPERATURE),
+                    )
+                    for hvac_id in find_all(
+                        "$..terminals[*].served_by_heating_ventilating_air_conditioning_system",
+                        zone,
+                    )
+                }
+
+        # find preheat_system's hot water loop type
+        hot_water_loop_type_dict = {}
+        for preheat_system in find_all("$..preheat_system", rmi_b):
+            hot_water_loop = preheat_system["hot_water_loop"]
+            hot_water_loop_type_dict[hot_water_loop] = find_exactly_one_fluid_loop(
+                rmi_b, hot_water_loop
+            )["type"]
+
+        # find applicable hvac sys ids
+        baseline_system_types_dict = get_baseline_system_types(rmi_b)
+        applicable_hvac_sys_ids = [
+            hvac_id
+            for sys_type in baseline_system_types_dict.keys()
+            for target_sys_type in APPLICABLE_SYS_TYPES
+            if baseline_system_type_compare(sys_type, target_sys_type, False)
+            for hvac_id in baseline_system_types_dict[sys_type]
+        ]
+
+        return {
+            "hvac_max_zone_setpoint_dict": hvac_max_zone_setpoint_dict,
+            "hot_water_loop_type_dict": hot_water_loop_type_dict,
+            "applicable_hvac_sys_ids": applicable_hvac_sys_ids,
+        }
+
+    def list_filter(self, context_item, data):
+        applicable_hvac_sys_ids = data["applicable_hvac_sys_ids"]
```

Comment:
I prefer creating a variable for the context_item as the first line of code inside a list_filter:
```
hvac_sys_b = context_item.baseline
```

Reply:
Fixed.